lsp/client/vscode/src/credentialsActivation.ts

```diff
@@ -0,0 +1,154 @@
+    }
+    stream.write(JSON.stringify(payload))
+    stream.write('\n')
+}
+
+/**
+ * Updates the language client's initialization payload to indicate that it can provide credentials
+ * for AWS language servers.
+ */
+export function configureCredentialsCapabilities(clientOptions: LanguageClientOptions) {
+    if (!clientOptions.initializationOptions) {
+        clientOptions.initializationOptions = {}
+    }
+
+    // This is how we configure the behavior of AWS Language Servers.
+    // The structure needs to be formalized across all AWS hosts/extensions.
+    //
+    // This structure is exploration/conceptual/speculative at this time.
+    // See lsp\core\aws-lsp-core\src\initialization\awsInitializationOptions.ts
+    clientOptions.initializationOptions.credentials = {
+        providesIam: true,
+    }
+}
+
+export async function registerIamCredentialsProviderSupport(
+    languageClient: LanguageClient,
+    extensionContext: ExtensionContext
+): Promise<void> {
+    extensionContext.subscriptions.push(
+        ...[
+            commands.registerCommand('awslsp.selectProfile', createSelectProfileCommand(languageClient)),
+            commands.registerCommand('awslsp.clearProfile', createClearProfileCommand(languageClient)),
+        ]
+    )
+}
+
+/**
+ * This command simulates an extension's credentials state changing, and pushing updated
+ * credentials to the server.
+ *
+ * In this simulation, the user is asked for a profile name. That profile's credentials are
+ * resolved and sent. (basic profile types only in this proof of concept)
+ */
+function createSelectProfileCommand(languageClient: LanguageClient) {
+    return async () => {
+        const profileName = await window.showInputBox({
+            prompt: 'Which credentials profile should the language server use?',
+        })
+
+        // PROOF OF CONCEPT
+        // We will resolve the default profile from the local system.
+        // In a product, the host extension would know which profile it is configured to provide to the language server.
+        const awsCredentials = await fromIni({
+            profile: profileName,
+        })()
+
+        const payload = createUpdateIamCredentialsPayload(awsCredentials)
+        await sendIamCredentialsUpdate(payload, languageClient)
+
+        languageClient.info(`Client: The language server is now using credentials profile: ${profileName}`)
+    }
+}
+
+/**
+ * Creates a response payload that contains encrypted data
+ */
+function createUpdateIamCredentialsPayload(awsCredentials: AwsCredentialIdentity): UpdateCredentialsPayload {
+    const responseData: UpdateIamCredentialsPayloadData = {
+        accessKeyId: awsCredentials.accessKeyId,
+        secretAccessKey: awsCredentials.secretAccessKey,
+        sessionToken: awsCredentials.sessionToken,
+    }
+
+    // encrypt payload, create response
+    const iv = crypto.randomBytes(16)
+    const encoder = crypto.createCipheriv('aes-256-gcm', encryptionKey, iv, {
```

Comment:
question:  Does everyone agree to use aes-256-gcm upfront?  Are there any one way doors when that algo is deprecated and it needs to change?

Reply:
That is still TBD, but is the current proposal. This is simply conceptual code that shows a working version.

Let's say that the algo needs to change some time in the future. New (semver-major) versions of the language server(s) would be released, using the new algo. New versions of the Toolkits would also be released, which support the new algo and consume the new major versions of the language servers. Previously released servers/Toolkits would remain compatible with each other, and continue using the old algo.